include/nbl/video/IGPUObjectFromAssetConverter.h

```diff
@@ -591,76 +710,189 @@ auto IGPUObjectFromAssetConverter::create(const asset::ICPUMesh** const _begin,
+    auto* signalFence = _params.transferFence ? fence.get() : nullptr;
+    _params.transferQueue->submit(1u, nullptr, signalFence);
+    if (_params.transferFence && fence)
+        _params.transferFence[0] = std::move(fence);
+    if (_params.transferSemaphore)
+        _params.transferSemaphore[0] = imgsFilledSemaphore;
+
+    if (!needToGenMips)
+        return res;
+
+    // TODO when mipmaps are generated with compute -- another submit for compute queue, wait on imgsFilledSemaphore
+    // now this code below instead:
+    // @matt How to synchronize mipmap generation and filling image storage??
+    for (ptrdiff_t i = 0u; i < assetCount; ++i)
+    {
+        auto& gpuimg = (*res)[i];
+        auto* cpuimg = _begin[i];
+
+        auto format = gpuimg->getCreationParameters().format;
+        if (!cpuimg->getRegions().size() || asset::isIntegerFormat(format))
+            continue;
+
+        uint32_t lowestPresentMip = 1u;
+        while (cpuimg->getRegions(lowestPresentMip).size()) // wtf??? @matt? how could it be working?
+            lowestPresentMip++;
+        // generate temporary image view to make sure we don't screw up any explicit mip levels
+        IGPUImageView::SCreationParams tmpViewParams;
+        tmpViewParams.subresourceRange.levelCount = cpuimg->getCreationParameters().mipLevels + 1u - lowestPresentMip;
+        // if not all mip levels have been manually specified
+        if (tmpViewParams.subresourceRange.levelCount > 1u)
+        {
+            tmpViewParams.flags = static_cast<IGPUImageView::E_CREATE_FLAGS>(0u);
+            tmpViewParams.image = gpuimg;
+            switch (cpuimg->getCreationParameters().type)
+            {
+            case asset::IImage::ET_1D:
+                tmpViewParams.viewType = IGPUImageView::ET_1D_ARRAY;
+                break;
+            case asset::IImage::ET_2D:
+                if (cpuimg->getCreationParameters().flags & asset::IImage::ECF_CUBE_COMPATIBLE_BIT)
+                    tmpViewParams.viewType = IGPUImageView::ET_CUBE_MAP_ARRAY;
+                else
+                    tmpViewParams.viewType = IGPUImageView::ET_2D_ARRAY;
+                break;
+            case asset::IImage::ET_3D:
+                tmpViewParams.viewType = IGPUImageView::ET_3D;
+                break;
+            default:
+                assert(false);
+                break;
+            }
+            tmpViewParams.format = cpuimg->getCreationParameters().format;
+            //tmpViewParams.subresourceRange.aspectMask
+            tmpViewParams.subresourceRange.baseMipLevel = lowestPresentMip - 1u;
+            tmpViewParams.subresourceRange.layerCount = cpuimg->getCreationParameters().arrayLayers;
+            auto tmpView = _params.device->createGPUImageView(std::move(tmpViewParams));
+
+            // deprecated OpenGL path (do with compute shader in the future)
+            _params.device->regenerateMipLevels(tmpView.get());
```

Comment:
TODO (in a week)

at least maybe make the `regenerateMipLevels` a method of the commandbuffer, so its easier to get the logic right rather than having to debug both the compute shader and the synchro at the same time in the future ?